src/libostree/ostree-sysroot.c

```diff
@@ -1586,6 +1586,9 @@ ostree_sysroot_init_osname (OstreeSysroot       *self,
  * specified, then no cleanup will be performed after adding the
  * deployment. Make sure to call ostree_sysroot_cleanup() sometime
  * later, instead.
+ *
+ * If %OSTREE_SYSROOT_SIMPLE_WRITE_DEPLOYMENT_FLAGS_RETAIN_PREVIOUS_VERSION is
+ * specified, then the previous version will not be garbage collected.
```

Comment:
OK, so testing this now, I don't think this is quite what we want. The case I'm trying to capture is this:

Imagine we have two deployments:
- `31.0` (booted)
- `30.0`

What I want to prevent here is `30.0` getting blown away if we `rpm-ostree install foobar`. But e.g. if we `rpm-ostree upgrade` and get `31.1`, then regular GC rules apply. But the logic below doesn't GC it.

Essentially, I think we just want `RETAIN_PREVIOUS_VERSION` to mean "keep at least 2 versions around". So I think we need to tweak the logic so that we just do this: get the version from `new_deployment`, and make sure that the first deployment that has a different version is retained. If it would've been retained anyway by existing rules, then the condition is also satisfied.

Does that make sense?

Reply:
Yes. With the updated version, my tests seem to reflect the intended behaviour you described here.